Calculate MD5 per account

aayushchouhan09 opened the pull request:
### Explain the changes
1.  Added a flag to calculate MD5 etag per account.

- [X] Tests added

utkarsh-pro commented on src/util/chunk_fs.js:
```diff
@@ -19,14 +19,15 @@ class ChunkFS extends stream.Transform {
         this.q_buffers = [];
         this.q_size = 0;
-        this.MD5Async = config.NSFS_CALCULATE_MD5 ? new (nb_native().crypto.MD5Async)() : undefined;
+        this.MD5Async = md5_enabled ? new (nb_native().crypto.MD5Async)() : undefined;
```
Can we have `this.MD5Async = (md5_enabled || config.NSFS_CALCULATE_MD5) ? new (nb_native().crypto.MD5Async)() : undefined;` so that we can preserve the older behaviour? Or preserving the older behaviour not a requirement?

achouhan09 replied:
No, Actually our requirement is to remove `this.NSFS_CALCULATE_MD5` which is a global flag and replace it with our new flag `force_md5_enabled`

utkarsh-pro replied:
Then I think we need to update the tests as well which might be setting this to true/false and then performing tests. Also if the intention is to get rid of `NSFS_CALCULATE_MD5` then maybe we can get rid of it from `config.js` as well?

achouhan09 commented on src/sdk/namespace_nb.js:
```diff
@@ -129,6 +129,7 @@ class NamespaceNB {
             active_triggers,
             operation
         });
+        params = _.omit(params, ['force_md5_etag_bucket', 'force_md5_etag_account']);
```
Yes correct, currently we are supporting this feature for nsfs only.

dannyzaken commented on src/sdk/object_sdk.js:
```diff
@@ -689,6 +698,8 @@ class ObjectSDK {
                 const ns = await this._get_bucket_namespace(params.bucket);
                 this._check_is_readonly_namespace(ns);
                 if (params.copy_source) await this.fix_copy_source_params(params, ns);
+                params.force_md5_etag_bucket = await this.read_force_md5_etag_for_bucket(params.bucket);
+                params.force_md5_etag_account = await this.read_force_md5_etag_for_account();
```
I'm not sure there is a reason not to send a single force_md5 flag, WDYT?
```suggestion
                params.force_md5_etag = await this.read_force_md5_etag_for_bucket(params.bucket) || await this.read_force_md5_etag_for_account();
```

achouhan09 replied:
Yes, that would be better I think. We can remove some unecessary checks in that way.

dannyzaken commented on src/sdk/object_sdk.js:
```diff
@@ -160,6 +160,15 @@ class ObjectSDK {
+        return bucket.force_md5_etag;
+    }
+
+    async read_force_md5_etag_for_account() {
```
There is no need to make this function `async`, it only accesses memory

achouhan09 replied:
yes

Commit:
Calculate MD5 per account

- Implemented force_md5_etag flag in account
- Added unit test for force_md5_etag in account

Signed-off-by: Aayush <aayush@li-2392f9cc-287a-11b2-a85c-9fcc04b05da6.ibm.com>

utkarsh-pro commented on src/test/unit_tests/test_system_servers.js:
```diff
@@ -193,6 +193,33 @@ mocha.describe('system_servers', function() {
+            email: EMAIL1,
+        });
+        assert(info.force_md5_etag === true);
+
```
Maybe we can also test if after setting this true NSFS really enforces MD5 calculation? Will be helpful to avoid regressions when someone else is working on this area.

achouhan09 replied:
Thanks, yes your are right.

We are not testing the complete flow here, this test is only to check the account part. I think, we can have one more test to check the whole flow of md5 calculation but for this PR the above test should be fine. @dannyzaken wdyt??

dannyzaken replied:
in the system servers test we mainly tests the APIs. in this case the force_md5 flag is also not having much effect, since it is only used for NSFS flows.
I agree that we should have separate tests for the functionality of the md5 calculations in NSFS

Commit:
Updated tests for md5_etag_calculation flow

Signed-off-by: Aayush <aayush@li-2392f9cc-287a-11b2-a85c-9fcc04b05da6.ibm.com>
